include/ddsrouter/participant/implementations/rtps/impl/DiscoveryServerParticipant.ipp

```diff
@@ -353,32 +338,92 @@ void DiscoveryServerParticipant<ConfigurationType>::enable_tls(
     descriptor->tls_config.add_verify_mode(
         eprosima::fastdds::rtps::TCPTransportDescriptor::TLSConfig::TLSVerifyMode::VERIFY_PEER);
 
-    if (client_only)
+    if (client)
     {
-        // Fail verification if the server has no certificate
-        descriptor->tls_config.add_verify_mode(
-            eprosima::fastdds::rtps::TCPTransportDescriptor::TLSConfig::TLSVerifyMode::VERIFY_FAIL_IF_NO_PEER_CERT);
+        if (!tls_configuration->can_be_client())
+        {
+            logError(DDSROUTER_DISCOVERYSERVER_PARTICIPANT,
+                    "TLS Configuration expected a Client configuration.");
+            throw ConfigurationException("TLS Configuration expected a Client configuration.");
+        }
+        else
+        {
+            enable_tls_client(descriptor, tls_configuration);
+        }
     }
     else
     {
-        // Password
-        descriptor->tls_config.password = get_param(tls_config, TLS_PASSWORD_TAG, false);
-        // Private key
-        descriptor->tls_config.private_key_file = get_param(tls_config, TLS_PRIVATE_KEY_TAG);
-        // DDS-Router certificate
-        descriptor->tls_config.cert_chain_file = get_param(tls_config, TLS_CERT_TAG);
-        // DH
-        descriptor->tls_config.tmp_dh_file = get_param(tls_config, TLS_DHPARAMS_TAG);
+        if (!tls_configuration->can_be_server())
+        {
+            logError(DDSROUTER_DISCOVERYSERVER_PARTICIPANT,
+                    "TLS Configuration expected a Server configuration.");
+            throw ConfigurationException("TLS Configuration expected a Server configuration.");
+        }
+        else
+        {
+            // Add configuration for server
+            enable_tls_server(descriptor, tls_configuration);
+
+            // In case it could also be client, add tls config
+            if (tls_configuration->can_be_client())
+            {
+                enable_tls_client(descriptor, tls_configuration);
+            }
+        }
     }
 
     logDebug(DDSROUTER_DISCOVERYSERVER_PARTICIPANT,
-            "TLS configured for Participant " << this->id() << ".");
+            "TLS configured.");
+}
+
+template <class ConfigurationType>
+void DiscoveryServerParticipant<ConfigurationType>::enable_tls_client(
+        std::shared_ptr<eprosima::fastdds::rtps::TCPTransportDescriptor> descriptor,
+        std::shared_ptr<security::TlsConfiguration> tls_configuration)
+{
+    std::shared_ptr<security::TlsConfigurationClient> tls_configuration_ =
+            std::dynamic_pointer_cast<security::TlsConfigurationClient>(tls_configuration);
+
+    if (!tls_configuration_)
+    {
+        utils::tsnh(
+            utils::Formatter() << "Error, client available TlsConfiguration does not cast to TlsConfigurationClient.");
+    }
+
+    // Fail verification if the server has no certificate
+    descriptor->tls_config.add_verify_mode(
+        eprosima::fastdds::rtps::TCPTransportDescriptor::TLSConfig::TLSVerifyMode::VERIFY_FAIL_IF_NO_PEER_CERT);
```

Comment:
I am not sure this should be done for the server+client case, that is why it used to be applied only in the `client_only` case.

Reply:
`only_client` remains to be used.